### ISSUE TITLE
Proposal: New ClockTimePicker Control

### ISSUE THREAD

**Issue by robloo** [microsoft/microsoft-ui-xaml#3605]
# Proposal: New ClockTimePicker Control

## Summary

Date and time pickers have some gaps in functionality illustrated as below:

 * DatePicker : Used to set a date by inputting individual components
 * CalendardDatePicker : Used to set a date by visually selecting its representation on a calendar
 * TimePicker : used to set a time by inputting individual components
 * **ClockTimePicker** : Proposed new control used to set a time by visually selecting its representation on an analog clock
 * DateTimePicker : A proposed control combining date/time pickers together.

## Rationale

A ClockTimePicker is available on several other platforms and allows fast selection of times in a visual way -- whether that be 24H or 12H clocks. This should be added to WinUI to nicely round out the feature set:

* Logically completes the date/time pickers enabling "visual" selection of times just like dates
* Enables faster and more intuitive selection of times
* Positions WinUI to be more competitive with other platforms that already have this type of control.

## Scope

| Capability | Priority |
| :---------- | :------- |
| Show a visual analog clock allowing for both hours/minutes to be selected by moving the clock hands | Must |
| Support switching between 12H and 24H times | Should |
| Support second selection | Could |
| Support time-zone selection | Could |

**Comment by prayaas-a** [comment]
Only if done right. I know many people find the analog time picker control on Android to be clunky and confusing.

**Comment by mdtauk** [comment]
I don't think displaying an analog clock is a great idea, but some kind of circular time picker, similar to Android is not out of the question.

Back in the Windows Phone days, I designed an app called MetroTime, and we included a circular picker, before the same kind of control showed up in Windows 8.

![image](https://user-images.githubusercontent.com/7389110/99042021-e5b94580-2583-11eb-8608-64a763a1396b.png)
_WP7 MetroTime App_

![image](https://user-images.githubusercontent.com/7389110/99042056-f964ac00-2583-11eb-8559-4492112545b7.png)
_Win8 Alarms App_

![image](https://user-images.githubusercontent.com/7389110/99042102-1305f380-2584-11eb-85f8-293e5de7fa87.png)
_Unused design for Win10 Mobile version of MetroTime_

![image](https://user-images.githubusercontent.com/7389110/99042299-67a96e80-2584-11eb-9b5c-55805d2ccd1f.png)
_Android "analog" Time Picker control_

**Comment by robloo** [comment]
It certainly doesn't have to look like the analog clock on Android. The main point is hours/minutes can be selected quickly because all values are visible at once due to the circular arrangement being more efficient with available space. @mdtauk Has some good pictures there. My only suggestion would be to find a way to show the other numerical values all around the circle at least during selection.

**Comment by mdtauk** [comment]
> It certainly doesn't have to look like the analog clock on Android. The main point is hours/minutes can be selected quickly because all values are visible at once due to the circular arrangement being more efficient with available space. @mdtauk Has some good pictures there. My only suggestion would be to find a way to show the other numerical values all around the circle at least during selection.
>

A lot of thought will need to go into this control.
- How will it work with HoloLens?
- How does it work on Xbox?
- How would Compact density impact this control? (This one may require WinUI 3 as flyouts can not currently change based on density) #925
- How does it adapt between Touch and Pointer invocation? Some ContextMenus currently increase touch targets when invoked by a touch gesture.
- How do you differentiate with guidance, when to use this control, versus the current Time Picker, or the proposed more keyboard friendly Time and Date Pickers? #918 #2255 #2380

**Comment by robloo** [comment]
> How will it work with HoloLens?

No experience with HoloLens so have no idea.

> How does it work on Xbox?

Those XBox control analog sticks with 360-degree input would actually be quite useful to quickly choose a position on a circle.

> How would Compact density impact this control? (This one may require WinUI 3 as flyouts can not currently change based on density)

Should do what the CalendarDatePicker does, if anything. These controls are more physicalized and should probably be desiged for touch from the start.

> How does it adapt between Touch and Pointer invocation? Some ContextMenus currently increase touch targets when invoked by a touch gesture.

Don't thing there is really a need to differentiate the two on such a proposed control. The hours/minutes circles will be wide enough for touch and pointer input most likely. Depends on the "final" design though.

> How do you differentiate with guidance, when to use this control, versus the current Time Picker, or the proposed more keyboard friendly Time and Date Pickers?

Even the CalendarDatePicker is stretching things to try to come up with [guidance](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/calendar-date-picker#is-this-the-right-control). It comes down to your particular use case and if visual input is useful for the end-user. For form-entry data oriented apps, use the TimePicker. For event planning apps, use a visual clock-picker.

**Comment by mdtauk** [comment]
> > How does it work on Xbox?
> 
> Those XBox control analog sticks with 360-degree input would actually be quite useful to quickly choose a position on a circle.
> 

Good Call, wonder if you should switch between hours, and minutes with a button, or allow the two analog sticks to control each separately?



> > How does it adapt between Touch and Pointer invocation? Some ContextMenus currently increase touch targets when invoked by a touch gesture.
> 
> Don't think there is really a need to differentiate the two on such a proposed control. The hours/minutes circles will be wide enough for touch and pointer input most likely. Depends on the "final" design though.
> 

The Radius of the radial selections needs to be large enough in touch so you can drag outside of the circle, and have finer precision picking.  Then you need to consider how that would work with Light Dismiss behaviours on the flyout containing the control.

With a mouse, it will need to be small enough to not feel "obnoxious" but still large enough to enable fine tuned "ticks" around the diameter, 

> > How do you differentiate with guidance, when to use this control, versus the current Time Picker, or the proposed more keyboard friendly Time and Date Pickers?
> 
> Even the CalendarDatePicker is stretching things to try to come up with [guidance](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/calendar-date-picker#is-this-the-right-control). It comes down to your particular use case and if visual input is useful for the end-user. For form-entry data oriented apps, use the TimePicker. For event planning apps, use a visual clock-picker.
>

This question will need the most attention if you are to persuade the teams that not only is there a need for this kind of control, but also that it would feel familiar from other platforms - fit in among other controls - and identify examples of current app, inbox, or shell experiences, that would benefit greatly from the addition of this control.

----

Finally, does this become part of the existing (or changed) Time Picker, or should it be it's own discrete control?

**Comment by robloo** [comment]
@mdtauk Yes, there is definitely more that needs to be planned, discovered and discussed, I was just hoping to start the discussion here. For this one unfortunately I don't have a clear design in my head -- yet.

> Finally, does this become part of the existing (or changed) Time Picker, or should it be it's own discrete control?

My proposal is for a new control. Just as the DatePicker and CalendarDatePicker exist for different reasons, the TimePicker and ClockTimePicker would exist for similar use cases.

**Comment by R3voA3** [comment]
https://user-images.githubusercontent.com/7389110/99042299-67a96e80-2584-11eb-9b5c-55805d2ccd1f.png

This is one of the best designs. The main advantage over the current time picker we have is, that I can select any value for hours and minutes without scrolling. It's just so much faster and easier to use.

**Comment by robloo** [comment]
Still valuable IMO
